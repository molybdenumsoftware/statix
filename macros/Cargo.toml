[package]
edition.workspace = true
license.workspace = true
name = "macros"
version = "0.0.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
hex.workspace = true
proc-macro2.workspace = true
quote.workspace = true
sha2.workspace = true
syn.workspace = true

[lib]
proc-macro = true

[lints]
workspace = true
