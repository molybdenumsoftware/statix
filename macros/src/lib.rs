mod explain;
mod metadata;

use explain::generate_explain_impl;
use metadata::{generate_meta_impl, RawLintMeta};
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{parse_macro_input, Ident, ItemStruct};

fn generate_self_impl(struct_name: &Ident) -> TokenStream2 {
    quote! {
        impl #struct_name {
            pub fn new() -> Self {
                Self
            }
        }
    }
}

#[proc_macro_attribute]
pub fn lint(attr: TokenStream, item: TokenStream) -> TokenStream {
    let struct_item = parse_macro_input!(item as ItemStruct);
    let meta = parse_macro_input!(attr as RawLintMeta);

    let struct_name = &struct_item.ident;
    let self_impl = generate_self_impl(struct_name);
    let meta_impl = generate_meta_impl(struct_name, &meta);
    let explain_impl = generate_explain_impl(&struct_item);

    (quote! {
        #struct_item

        ::lazy_static::lazy_static! {
            pub static ref LINT: Box<dyn crate::Lint> = Box::new(#struct_name::new());
        }

        #self_impl
        #meta_impl
        #explain_impl

        impl crate::Lint for #struct_name {}
    })
    .into()
}
