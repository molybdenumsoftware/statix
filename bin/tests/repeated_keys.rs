mod _utils;

use indoc::indoc;

use macros::generate_tests;

generate_tests! {
    rule: useless_has_attr,
    expressions: [
        // fine
        "{ foo.bar = 1; }",

        // do not raise on rec
        indoc! {"
            rec {
              foo.x = foo.y;
              foo.y = 2;
              foo.z = 3;
            }
        "},

        // exactly 3 occurrences
        indoc! {r#"
            {
              foo.bar = 1;
              foo.bar."hello" = 1;
              foo.again = 1;
            }
        "#},

        // more than 3, omit the extra
        indoc! {"
            {
              foo.baz.bar1 = 1;
              foo.baz.bar2 = 2;
              foo.baz.bar3 = 3;
              foo.baz.bar4 = 4;
              foo.baz.bar5 = 5;
            }
        "},
    ],
}
