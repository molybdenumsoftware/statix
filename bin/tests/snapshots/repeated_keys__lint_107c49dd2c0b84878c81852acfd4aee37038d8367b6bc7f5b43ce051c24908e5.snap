---
source: bin/tests/repeated_keys.rs
expression: "\"{\\n  foo.bar = 1;\\n  foo.bar.\\\"hello\\\" = 1;\\n  foo.again = 1;\\n}\\n\""
---
[W20] Warning: Avoid repeated keys in attribute sets
   ╭─[<temp_file_path>:2:3]
   │
 2 │   foo.bar = 1;
   ·   ───┬───  
   ·      ╰───── The key foo is first assigned here ...
 3 │   foo.bar."hello" = 1;
   ·   ───────┬───────  
   ·          ╰───────── ... repeated here ...
 4 │   foo.again = 1;
   ·   ────┬────  
   ·       ╰────── ... and here. Try foo = { bar=...; bar."hello"=...; again=...; } instead.
───╯
